sunxspex/sunxspex_fitting/stix_spec_code.py

```diff
@@ -2,113 +2,117 @@
-
-    channel_bins_inds, channel_bins = _return_masked_bins(sdict)
-
-    lvt = np.ones((len(t_lo), len(channel_bins)))  # livetimes, (rows,columns) -> (times, channels), for STIX it is 1 at the minute?
-
-    # get counts [counts], count rate [counts/s], and error on count and count rate
-    counts, counts_err, cts_rates, cts_rate_err = _spec_file_units_check(stix_dict=sdict, time_dels=time_deltas)
-
-    return channel_bins, channel_bins_inds, time_bins, lvt, counts, counts_err, cts_rates, cts_rate_err
-
-
-def _return_masked_bins(sdict):
-    """ Return the energy bins where there is data.
-
-    Parameters
-    ----------
-    sdict : dict
-            Dictionary containing all STIX spectral file information.
-
-    Returns
-    -------
-    A 2d array of the energy bin edges.
-    """
-    # get all energy bin edges (0--inf keV)
-    e_bins = np.concatenate((sdict["4"][1]['e_low'][:, None], sdict["4"][1]['e_high'][:, None]), axis=1)
-
-    # get all indices of the energy bins needed
-    mask_inds = sdict["1"][1]['energy_bin_edge_mask'][0].astype(bool)
-
-    return mask_inds, e_bins
-
-
-def _spec_file_units_check(stix_dict, time_dels):
-    """ Make sure STIX count data is in the correct units.
-
-    This file is regularly saved out using different units.
-
-    Parameters
-    ----------
-    stix_dict : dict
-            Dictionary containing all STIX spectral file information.
-
-    time_dels : 1d array
-            The time duration of each recorded spectrum.
-
-    Returns
-    -------
-    A 2d array of the counts [counts], count rates [counts/s], and the count and count
-    rate errors (counts, counts_err, cts_rates, cts_rate_err).
-    """
-    # stix can be saved out with counts, counts/sec, or counts/sec/cm^2/keV using counts, rate, or flux, respectively
-    if stix_dict["0"][0]["BUNIT"] == "counts":
-        counts = stix_dict["2"][1]["counts"][:, :]
-        counts_err = np.sqrt(counts)  # should this be added in quadrature to the estimated count compression error
-        cts_rates = counts / time_dels[:, None]
-        cts_rate_err = counts_err / time_dels[:, None]
-    else:
-        print("I don\'t know what units STIX has.")
-
-    return counts, counts_err, cts_rates, cts_rate_err
-
-
-def _get_srm_file_info(srm_file):
+    with fits.open(spec_file) as sf:
+        # get units from the fits columns rather than hard-coding
+        slice_unit = lambda hdu, c: u.Unit(sf[hdu].columns[c].unit)
+
+        time_mids = sf['data'].data['time'] << slice_unit('data', 'time')
+        dt = sf['data'].data['timedel'] << slice_unit('data', 'timedel')
+        # if odd `t` in seconds then edges are [midt-floor(bin_width/2), midt+ceil(bin_width/2)]
+        # e.g., mid_t=19, del_t=13 then edges would be [19-6, 19+7]=[13,26]
+        t_lo = time_mids - (np.floor(dt.value / 2) << dt.unit)
+        t_hi = time_mids + (np.ceil(dt.value / 2) << dt.unit)
```

Comment:
Not sure the floor / ceil are correct, take a time of 50, and tiemdel of 13 the edges should be `50 - 13/2`, `50 + 13/2` so 43.5, 56.5 not 50 - 6, and 50+7 ?

Reply:
This was the case about over a year ago when I wrote the code for some of the first STIX files I got to load into Sunxspex. It's because the time bins are contiguous and the file only gave the "mid-time" and the bin-width, both given as integers of centi-seconds, so the bin edges couldn't be at 9.5 (or anything.5) unless the first time bin started at 0.5 s or all `timedels` are odd values. It definitely needs check for the files now just to be safe since `time_mids` might not be given as integers anymore.